Improve packaging for lsif-clang by bundling dynamic libraries

varungandhi-src opened the pull request:
Based off #78.

(cherry picked from commit 4d1b5d95e9da837f0fccabfd56f7c8ee4bb848f9)

### Test plan

- [x] Add CI step to build and test this image

We can add a publish step in a follow-up PR if needed.

varungandhi-src commented on ubuntu_1804.Dockerfile:
```diff
@@ -0,0 +1,34 @@
+FROM ubuntu:18.04 as build
```
(Copying my comment from earlier in https://github.com/sourcegraph/lsif-clang/pull/78#discussion_r914640837)

> 

Question: Is there a specific reason for using Ubuntu 18.04? If we don't have code sharing between this Dockerfile and the top-level one, it'd be nice to at least have the same exact code being duplicated across this Dockerfile and the main top-level Dockerfile.
>
> For example, another PR just added libdwarf-dev and libelf-dev as dev dependencies. It wouldn't be too much work to add them here too, and I don't expect we'll be adding many more of those, but having the same exact code sequence would make changes smoother.

cc @jlisee

varungandhi-src commented on clang-tools-extra/lsif-clang/package/copy_dependencies.sh:
```diff
@@ -0,0 +1,15 @@
+binary="$1"
+output="$2"
+
+for lib in $(ldd "$binary" | cut -d ">" -f 2 | awk '{print $1}' | grep -vE "(linux-vdso\.|ld-linux-x86-64\.|libc\.)"); do
+    cp --verbose "$lib" "$output/"
+done
```
We should double-check if the new `libdwarf-dev` and `libelf-dev` dependencies require updating this.

varungandhi-src replied:
Looks like this needed an update for `librt` and `libpthread`; copying them across OSes lead to errors, but removing them fixes the issue. See the commit message of be6a3a11aa2636073b6676b845d0d917afa4d32e for details.

But it works fine otherwise.

varungandhi-src commented:
When running the installation script 

```
./clang-tools-extra/lsif-clang/package/build.sh
```

inside an Ubuntu 20.04 VM, I get an error with missing packages

```
WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package llvm-11
E: Unable to locate package clang-11
E: Unable to locate package libclang-11-dev
The command '/bin/sh -c apt install -y llvm-11 clang-11 libclang-11-dev cmake' returned a non-zero code: 100
```

This happens both with the original commit and after the refactoring a392f4dd0545db286630bb08eea710edd4023b56.

jlisee commented on Bundled.Dockerfile:
```diff
@@ -0,0 +1,31 @@
+#
+# See issue https://github.com/sourcegraph/lsif-clang/issues/72
+
+FROM ubuntu:22.04 as build
```
See my note, building on such a new version of Ubuntu greatly restricts the platforms that can run the resulting release.

jlisee reviewed (commented):
glibc has a high degree of backwards compatibility so the if we build on Ubuntu 18.04 and bring all the other libraries we should be able to run on a large number of linux systems of that age and newer.  This should allow using `lsif-clang` in most environments that people are building C++.

To share code between Dockerfile's this is what I would do:

 - Create a `clang-tools-extra/lsif-clang/install_dependencies.sh` which uses `lsb_release` to be be portable and does the cmake, llvm and clang apt based installs
 - Create a `clang-tools-extra/lsif-clang/configure.sh` which does the `CC=... CXX=... cmake` steps

jlisee commented:
> Hmm, this doesn't seem to work for me across Ubuntu versions, because of a glibc version mismatch. Maybe because of the upgrade to Ubuntu 22.04 for the image?

Yes sorry I didn't replay last night, but the entire point of using 18.04 is that you can still build new software on it but the resulting binaries are widely portable.

varungandhi-src commented:
No worries about the response, I think you're in a US timezone, whereas I am in Taipei right now. I'd like to double-check the change manually once before merging. On aarch64, it doesn't work, so I'm going to test with x86_64.

As for debugging, I've been working on updating the Python driver script so that it can perform index merging after-the-fact, gracefully handling OOMs since it is multi-process instead of multi-threaded. I'm not 100% sure if it'll work (without blowing up the index size) but just giving you a heads up to avoid duplicate work.

varungandhi-src commented:
I've reverted the bump to 22.04 for the Dockerfile. There were also some other bugs that I noticed:

- I've renamed the file and cleaned some things.
- The `cd` in the shim bash script seemed to lead to incorrect paths for `lsif-clang`, so it would look in the wrong place for the `compile_commands.json` and fail to find it (not 100% sure if my "refactoring" broke this or if the `cd` earlier was also incorrect). So I've removed that.
- I noticed linker errors when copying `librt` and `libpthread` across Ubuntu versions (from 18.04 to 20.04), so I've removed those.

Thank you for submitting the original PR with this approach. I was investigating rewriting the CMake in the style of upstream Clang-based tools, which are often statically linked in a busybox-style binary. This approach feels hackier, but if it gets the job done for you, I don't have complaints.

jlisee reviewed (approved):
I was able to successfully to run this locally on my x64, Ubuntu 18.04 machine.

varungandhi-src commented on Bundled_Ubuntu1804.Dockerfile:
```diff
@@ -0,0 +1,55 @@
+      | gpg --dearmor - \
+      | tee /etc/apt/trusted.gpg.d/kitware.gpg \
+      > /dev/null && \
+    echo "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-11 main" \
```
I think hard-coding the bionic is fine, as one would need to tweak the rest of the file for other distros anyways.
